nemo/collections/vlm/qwen2vl/model/base.py

```diff
@@ -0,0 +1,1096 @@
+
+            input_ids_text = input_ids.clone()
+            # MultiModal Token indices are assumed to be values
+            input_ids_text[input_ids_text < 0] = 0
+
+            language_embeddings = self.language_model.embedding(
+                input_ids=input_ids_text, position_ids=None
+            )  # [decoder_seq_len, b, h_language]
+
+            if self.sequence_parallel_lm:
+                # Gather the language embeddings back.
+                # We use the full embedding to insert image embeddings
+                # and then scatter to avoid load imbalance.
+                language_embeddings = gather_from_sequence_parallel_region(
+                    language_embeddings, tensor_parallel_output_grad=False
+                )
+                # Remove the padding done for SP as we'll need new padding calculation
+                # after image embeddings are inserted.
+                if padded_seq_len != 0:
+                    language_embeddings = language_embeddings[:-padded_seq_len]
+
+            language_embeddings = language_embeddings.transpose(1, 0).contiguous()  # [b, decoder_seq_len, h_language]
+
+        # Preprocess input, labels and loss mask.
+        combined_embeddings, final_labels, final_loss_mask = self._preprocess_data(
+            input_ids,
+            loss_mask=loss_mask,
+            labels=labels,
+            language_embeddings=language_embeddings,
+            image_embeddings=image_embeddings,
+            video_embeddings=video_embeddings,
+        )  # [decoder_seq_len, b, h_language], [b, decoder_seq_len], [b, decoder_seq_len]
+
+        if attention_mask is None:
+            final_attention_mask = attention_mask
+        else:
+            # TODO: use no cache opiton for now
+            past_seen_tokens = 0  # past_key_values.get_seq_length() if past_key_values is not None else 0
+            cache_position = torch.arange(
+                past_seen_tokens, past_seen_tokens + combined_embeddings.shape[0], device=combined_embeddings.device
+            )
+
+            final_attention_mask = self._update_causal_mask(
+                attention_mask,
+                combined_embeddings,
+                cache_position,
+            )
+
+        output = self.language_model(
+            input_ids=None,
+            position_ids=position_ids,
+            attention_mask=final_attention_mask,
+            decoder_input=combined_embeddings,
+            labels=final_labels,
+            inference_params=inference_params,
+            runtime_gather_output=runtime_gather_output,
+        )  # output shape: [batch_size, seq length, vocab_size]
+
+        if labels is None or loss_mask is None:
+            return output
+        else:
+            return output, final_loss_mask.contiguous()
+
+    # override _preprocess_data() in megatron-lm/megatron/core/models/multimodal/llava_model.py
+    def _preprocess_data(
```

Comment:
we recently have updated this func in neva a bit. Can you double check to see if anything need to be updated.